### ISSUE TITLE
Commitment to Gitsync

### ISSUE THREAD

**Issue by aschrijver** [burtonator/polar-bookshelf#765]
**DONATE OR PURCHASE POLAR PREMIUM FOR PRIORITIZED SUPPORT AND FEATURES **

**DONATE OR PURCHASE POLAR PREMIUM FOR PRIORITIZED SUPPORT AND FEATURES **

https://opencollective.com/polar-bookshelf

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by burtonator** [comment]
Thanks @aschrijver ... 

Polar supports a 'Datastore' concept which is really just an interface so Polar supports a long term vision where 3rd parties could design their own datastores and tie them into Polar and just use that directly.

https://github.com/burtonator/polar-bookshelf/blob/6cc76820ab3da2116f0b80f608d5f83f28082734/web/js/datastore/Datastore.ts

I don't think there's anything specific about Firebase what's exposed past the FirebaseDatastore that would lock us down to Firebase.

It would just require a platform that was event-based, had some sort of auth layer, etc.

In theory one could probably use Elasticsearch with some sort of binary storage layer as well an event distribution framework.  But it would be a real pain to setup.

The reason I like Firebase actually has to do with the fact that it's amazingly easy to setup auth as well as the fact that it has a VERY VERY nice event listener framework.

It's real time which is something that git lacks. The longer you're out of sync the greater chance you need to merge.  With Firebase you're constantly in sync.  Even remote machines that are running in the background are sync'd in real time.  

Also, the main reason we're not really pushing git sync is mostly due to both ease of use as well as the fact that git sync isn't really a consumer-friendly tech.

The primary goals of Polar are to focus on mass consumer adoption of the collaboration aspects of Polar.  

By consumer I mean just regular people using Polar to collaborate with colleagues.  

I wrote up a post about it here:

https://getpolarized.io/2019/03/01/polar-personal-knowledge-repository.html

Any type of self-hosted strategy means that the user needs to configure software and it makes things much more complicated and would slow our growth.

I'm very motivated by the "humane technology" and "self-hosted" movement but my thinking is that this is economic and a byproduct of excessive capitalism not necessary something inherent in tech or Google/Facebook.

This is the main reason I'm trying to pursue more rapidly adoptable technologies.  For Polar to succeed long term I want the community and the technology to be economically viable and to do that I need to have a large user base.  

I think it can still be humane.  In fact that's why I called it Polar and made it Open Source.  I want people to trust that it will never vanish which is why the OSS part of it is very important.

I think our next steps might be to apply for some grants to continue funding Polar moving forward as I think we're still a bit away from the large scale user base that we're trying to get to ... would be good to get your advice next week if you're free.

**Comment by burtonator** [comment]
There are also some rough technology challenges here though. Features like full-text search, blob storage, auth, etc.  They're not easily solved on a self-hosted strategy yet.  If something like Elasticsearch supported blob storage out of the gate that might be interesting.

**Comment by aschrijver** [comment]
Wow, thank you for your elaborate response, @burtonator !

Having looked at Firebase for a past project, I share your enthusiasm for the technology and agree with all these points. It is very easy to add true power to an application by adding Firebase to the mix (though limitations may be found later, when wanting to support more complex features that cannot be found in Firebase services).

> Also, the main reason we're not really pushing git sync is mostly due to both ease of use as well as the fact that git sync isn't really a consumer-friendly tech.

This is also a requirement for the Humane Tech Community, as well as limiting the amount of tools we use. Currently we are mostly a Discourse forum an Github organization. With so much cool Markdown (and Asciidoc and others) OSS projects we can come a long way to make that the easy-to-edit for everyone default format, and having git repo's and Github be our backend storage, without exposing end-users to most of the complexity.

Note that in our case the more technical people will do the installation (but this could just be spinning up some Docker images, or similarly easy ways).

> Also, the main reason we're not really pushing git sync 

"Not really pushing" is no problem for us, but "Eventually may drop" may be problematic for our tool selection.

> My thinking is that this is economic and a byproduct of excessive capitalism not necessary something inherent in tech or Google/Facebook.

I understand how you mean this and agree on our forms of capitalism as root cause of the problem. but wouldn't saying it like this be similar to: "This suppression is is a byproduct of authoritarianism, but not inherent to Dictator XYZ"?

Big Tech is a big problem and it is growing. We don't see most of the consequences yet, but their domination has many harms of technology involved, as we regularly discussed on our forum. After the reorganization we are planning (building The Pyramids of Humane Technology), one of the credo's we will adopt is "Small Tech" (coined by [Aral Balkan](https://ar.al)). Not big tech, but small tech.

Google is the pinnacle of big tech, and the founder of surveillance capitalism. I want to avoid including their services in our community's tool belt :)

**Comment by burtonator** [comment]
OK.. to clarify here you're looking for a tool that  the Humane Tech Community can adopt correct? and you want to use git for collaboration?

> : "This suppression is is a byproduct of authoritarianism, but not inherent to Dictator XYZ"?

More like if there is no alternative to authoritarianism then we're going to keep having the same problems even if we replace dictator XYZ... 

> Google is the pinnacle of big tech, and the founder of surveillance capitalism. I want to avoid including their services in our community's tool belt :)

I think this is where we differ somewhat.  These companies are so massive it's hard to avoid them to a certain degree.  They're going to be somewhere in your stack but you should just try to minimize any downsides.  

Google and the other BigCos contribute to Linux + Docker + git for example but we don't just stop using them... 

I think one of the things I want with Polar is for it to be more of a force for good but more focused on what we're trying to accomplish - re-inventing education, scientific collaboration, and long term retention of knowledge.  

It's hard to serve two masters.  The Firebase thing isn't perfect but if it helps nail our MVP and get more adoption then seems like a win/win.

I'd like to see it grow more into a Wikimedia or Wordpress.org/com sort of situation where there is a non-profit designed to protect the community.

User growth is insanely insanely difficult. If you build an amazing tool you still have to shove it down peoples throats because it's very difficult to get user adoption.  

At least with Firebase + web + mobile I can hopefully build out the platform but still ship desktop+git sync because we're funded and have a large user base.

**Comment by aschrijver** [comment]
> OK.. to clarify here you're looking for a tool that the Humane Tech Community can adopt correct? and you want to use git for collaboration?

Yes, indeed. Given our stance on technology (adopting the ideas of [Aral Balkan's](https://ar.al) Small Tech vision), avoiding Big Tech that is at the root of many of the Harms of Technology and our community's solution-focus in [new Vision and Mission](https://community.humanetech.com/t/be-a-builder-help-improve-wellbeing-freedom-and-society/3322) (reorganization in the works) we will adopt mostly OSS without the links to FAANG. And we minimize tools, so no Google / Google Firebase accounts, etc.

> I think this is where we differ somewhat. These companies are so massive it's hard to avoid them to a certain degree. They're going to be somewhere in your stack but you should just try to minimize any downsides.
>
> Google and the other BigCos contribute to Linux + Docker + git for example but we don't just stop using them...

They do contribute a lot of OSS indeed. But they also _take a lot_ from OSS community. You are probably in the loop of the existential crisis in F(L)OSS realms, where maintainers of hugely successful projects find it really hard to eek out a living off their work, and tech giants making millions/billions adopting it.

As said I totally get your enthusiasm for Firebase. Would be great to see it go in direction of Wikimedia / WP direction. In the meantime it would, imho, be a great feature for Polar to be modularized, and maybe even have Gitsync and Firebase exist in separate projects, and allow anyone to add their own modules for persistence, etc.

Indeed user growth is very hard :smile: 

I know that from experience with Humane Tech Community, which is a vast, often abstract subject area. Your project rocks, so I am sure you will get ever more uptake. Current direction with Firebase makes sense. If you plan for modularity early (maybe you already do, but I didn't deep-dive into codes yet) you will tap into even more channels to growth.

**Comment by burtonator** [comment]
> You are probably in the loop of the existential crisis in F(L)OSS realms, where maintainers of hugely successful projects find it really hard to eek out a living off their work, and tech giants making millions/billions adopting it.

Yes... this is the problem I was discussing before where it's economic and I'm very very pessimistic about a fix until we solve the root economic problem.

> As said I totally get your enthusiasm for Firebase. Would be great to see it go in direction of Wikimedia / WP direction. In the meantime it would, imho, be a great feature for Polar to be modularized, and maybe even have Gitsync and Firebase exist in separate projects, and allow anyone to add their own modules for persistence, etc.

Unfortunately, due to technical reasons I don't think we'll have a plugin API in the short term but wouldn't turn away a PR as long as the code is high quality.  

The current plan for growth is probably focus on expanding the user base and then maybe applying for grants .
